lang/src/org/partiql/lang/eval/visitors/StaticTypeInferenceVisitorTransform.kt

```diff
@@ -530,120 +529,74 @@ internal class StaticTypeInferenceVisitorTransform(
+                return StaticType.INT
             }
 
-            return when (aggFunc.text) {
-                // current implementation of count will always return a long
-                "count" -> StaticType.INT8
-                // max/min supports all type and the result depends on comparison.
-                // aggregate function will not return missing as a potential Type
-                // in case that argType contains only missing,
-                "max", "min" -> {
-                    val possibleReturnTypes = argType.allTypes.filter {
-                        it !is MissingType
-                    }
-                    if (possibleReturnTypes.isEmpty())
-                        StaticType.NULL
-                    else
-                        StaticType.unionOf(possibleReturnTypes.toSet()).flatten()
+            val elementTypes = elementType.allTypes
+
+            fun List<StaticType>.convertMissingToNull() = toMutableSet().apply {
+                if (contains(StaticType.MISSING)) {
+                    remove(StaticType.MISSING)
+                    add(StaticType.NULL)
                 }
-                // current implementation of avg always return a decimal or null.
-                "avg" -> {
-                    when {
-                        !isNullOrMissingOrNumeric(argType) -> {
-                            handleInvalidInputTypeForAggFun(sourceLocation, aggFunc, argType, StaticType.unionOf(StaticType.MISSING, StaticType.NULL, StaticType.NUMERIC))
-                            StaticType.unionOf(StaticType.NULL, StaticType.DECIMAL)
-                        }
-                        // missing, null, or missing and null
-                        argType.isSubTypeOf(StaticType.unionOf(StaticType.MISSING, StaticType.NULL)) -> StaticType.NULL
-                        else -> StaticType.DECIMAL
+            }
+
+            fun StaticType.isUnknownOrNumeric() = isUnknown() || isNumeric()
+
+            return when (funcName) {
+                // In case that any element is MISSING or there is no element, we should return NULL
+                "max", "min" -> StaticType.unionOf(elementTypes.convertMissingToNull())
+                "sum" -> when {
+                    elementTypes.all { !it.isUnknownOrNumeric() } -> {
+                        handleInvalidInputTypeForAggFun(sourceLocation, funcName, elementType, StaticType.unionOf(StaticType.NULL_OR_MISSING, StaticType.NUMERIC).flatten())
+                        StaticType.unionOf(StaticType.NULL, StaticType.NUMERIC)
                     }
+                    elementTypes.any { !it.isUnknownOrNumeric() } -> StaticType.unionOf(
+                        elementTypes.filter { it.isUnknownOrNumeric() }.toMutableSet().apply { add(StaticType.MISSING) }
+                    )
+                    // In case that any element is MISSING or there is no element, we should return NULL
+                    else -> StaticType.unionOf(elementTypes.convertMissingToNull())
```

Comment:
I'm slightly confused by these two cases. What is the case on L578 describing and why is `MISSING` being added? From the comment on L571, you said the aggregation functions will return `NULL`. Is this for the case of a mistyped argument? If so, some of the comments may need adjusted.

Reply:
I believe a general principle the current implementation of type inferencer is that, if by any chance there could be a data type mismatch error, we should add `MISSING` to the result types to indicate it. e.g. the expression `a + b`, with binding of `"a" to INT, "b" to unionOf(INT, STRING)`, returns `unionOf(INT, MISSING)` instead of just `INT`.

Reply:
Ah ok that makes more sense, I think adding a comment would be helpful for understanding this case. Something like "non-numeric, non-unknown types will result in adding MISSING to return types"

Reply:
Will apply this comment